core/server/api/canary/users.js

```diff
@@ -8,6 +8,30 @@ const permissionsService = require('../../services/permissions');
 const ALLOWED_INCLUDES = ['count.posts', 'permissions', 'roles', 'roles.permissions'];
 const UNSAFE_ATTRS = ['status', 'roles'];
 
+function permissionOnlySelf(frame) {
+    const targetId = frame.options.id;
+    const userId = frame.user.id;
+    if (targetId !== userId && targetId !== 'me') {
+        return Promise.reject(new errors.NoPermissionError({message: i18n.t('errors.permissions.noPermissionToAction')}));
+    }
+    return Promise.resolve();
+}
+
+function getTargetId(frame) {
+    return frame.options.id === 'me' ? frame.user.id : frame.options.id;
+}
+
+async function fetchOrCreatePersonalToken(userId) {
+    const token = await models.ApiKey.findOne({user_id: userId}, {});
+
+    if (!token) {
+        const newToken = await models.ApiKey.forge({user_id: userId, type: 'admin'}).save();
```

Comment:
I think you should use the `add` method here if possible, we try to use add,findOne,edit,destroy methods rather than the lower level bookshelf methods